File: frontend/src/types/task.ts
export type TaskStatus =
  | 'pending'
  | 'queued'
  | 'running'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'skipped';

export type QuestionStatus = 'pending' | 'running' | 'completed' | 'failed';

export type AIProvider = 'gemini' | 'claude' | 'openai_compat';

export interface ProviderInfo {
  id: string;
  name: string;
  description: string;
}

export interface ApplicationQuestion {
  id: string;
  question: string;
  word_limit: number;
  answer?: string;
  status: QuestionStatus;
  error_message?: string;
  created_at: string;
  answered_at?: string;
}

export type TaskStep =
  | 'generate_resume'
  | 'compile_latex'
  | 'extract_text'
  | 'generate_cover_letter'
  | 'create_cover_pdf';

export interface StepProgress {
  step: TaskStep;
  status: TaskStatus;
  message: string;
  attempt: number;
  started_at?: string;
  completed_at?: string;
}

export interface Task {
  id: string;
  task_number: number;
  job_description: string;
  status: TaskStatus;
  generate_cover_letter: boolean;
  template_id: string;
  language?: string;
  experience_level?: string;
  provider?: string | null;
  pipeline_version?: string;
  steps: StepProgress[];
  resume_pdf_path?: string;
  cover_letter_pdf_path?: string;
  latex_source?: string;
  error_message?: string;
  validation_warnings?: string[];
  company_name?: string;
  position_name?: string;
  questions?: ApplicationQuestion[];
  cancelled?: boolean;
  created_at: string;
  completed_at?: string;
}

export interface ProgressUpdate {
  task_id: string;
  task_number: number;
  step: TaskStep;
  status: TaskStatus;
  message: string;
  attempt: number;
  node?: string;
  pipeline_version?: string;
}

export interface WebSocketMessage {
  type: 'progress';
  data: ProgressUpdate;
}

export interface Template {
  id: string;
  name: string;
  description: string;
}

export interface JDHistoryEntry {
  text: string;
  preview: string;
  saved_at: string;
}

export interface Toast {
  id: string;
  type: 'success' | 'error' | 'info';
  message: string;
}

export interface AppSettings {
  default_provider: string;
  gemini_api_key: string;
  gemini_model: string;
  gemini_temperature: number | null;
  gemini_top_k: number | null;
  gemini_top_p: number | null;
  gemini_max_output_tokens: number | null;
  gemini_thinking_level: string;
  gemini_enable_search: boolean;
  claude_api_key: string;
  claude_model: string;
  claude_temperature: number | null;
  claude_max_output_tokens: number | null;
  claude_extended_thinking: boolean;
  claude_thinking_budget: number;
  openai_compat_base_url: string;
  openai_compat_api_key: string;
  openai_compat_model: string;
  openai_compat_temperature: number | null;
  openai_compat_max_output_tokens: number | null;
  claude_proxy_base_url: string;
  claude_proxy_api_key: string;
  claude_proxy_model: string;
  claude_proxy_temperature: number | null;
  claude_proxy_max_output_tokens: number | null;
  deepseek_api_key: string;
  deepseek_model: string;
  deepseek_temperature: number | null;
  deepseek_max_output_tokens: number | null;
  qwen_api_key: string;
  qwen_model: string;
  qwen_temperature: number | null;
  qwen_max_output_tokens: number | null;
  enforce_resume_one_page: boolean;
  enforce_cover_letter_one_page: boolean;
  max_page_retry_attempts: number;
  generate_cover_letter: boolean;
  max_latex_retries: number;
  default_template_id: string;
  default_experience_level: string;
  allow_ai_fabrication: boolean;
  enable_contact_replacement: boolean;
  enable_text_validation: boolean;
  enable_llm_validation: boolean;
  user_linkedin_url: string;
  user_github_url: string;
  agent_providers: Record<string, string>;
}

export interface EvaluationData {
  ats_score: number;
  ats_breakdown: {
    keyword_similarity: number;
    semantic_similarity: number;
    skill_coverage: number;
    fuzzy_match: number;
    resume_quality: number;
    section_bonus: number;
    action_verbs_score: number;
    quantified_score: number;
    section_score: number;
    format_score: number;
  };
  matched_keywords: string[];
  missing_keywords: string[];
  combined_score: number;
  passed: boolean;
  llm_score?: number | null;
  llm_breakdown?: {
    keyword_alignment: number;
    professional_tone: number;
    quantified_achievements: number;
    relevance: number;
    ats_compliance: number;
    reasoning: string;
    improvements: string[];
  } | null;
}

export type Prompts = Record<string, string>;

File: backend/models/task.py
import uuid
from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class TaskStatus(StrEnum):
    PENDING = "pending"
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    SKIPPED = "skipped"


class QuestionStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ApplicationQuestion(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4())[:8])
    question: str
    word_limit: int = 150
    answer: str | None = None
    status: QuestionStatus = QuestionStatus.PENDING
    error_message: str | None = None
    created_at: datetime = Field(default_factory=datetime.now)
    answered_at: datetime | None = None


class TaskStep(StrEnum):
    GENERATE_RESUME = "generate_resume"
    COMPILE_LATEX = "compile_latex"
    EXTRACT_TEXT = "extract_text"
    GENERATE_COVER_LETTER = "generate_cover_letter"
    CREATE_COVER_PDF = "create_cover_pdf"


class StepProgress(BaseModel):
    step: TaskStep
    status: TaskStatus = TaskStatus.PENDING
    message: str = ""
    attempt: int = 0
    started_at: datetime | None = None
    completed_at: datetime | None = None


class Task(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4())[:8])
    task_number: int
    job_description: str
    status: TaskStatus = TaskStatus.PENDING
    generate_cover_letter: bool = True
    template_id: str = "classic"
    language: str = "en"
    provider: str | None = None
    pipeline_version: str = "v2"
    steps: list[StepProgress] = Field(default_factory=list)
    resume_pdf_path: str | None = None
    cover_letter_pdf_path: str | None = None
    cover_letter_text: str | None = None
    latex_source: str | None = None
    error_message: str | None = None
    cancelled: bool = False
    questions: list[ApplicationQuestion] = Field(default_factory=list)
    company_name: str = ""
    position_name: str = ""
    failed_latex_attempts: list[str] = Field(default_factory=list)
    agent_outputs: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.now)
    completed_at: datetime | None = None

    def model_post_init(self, __context):
        if not self.steps:
            self._rebuild_steps()

    def _rebuild_steps(self):
        self.steps = [
            StepProgress(step=TaskStep.GENERATE_RESUME),
            StepProgress(step=TaskStep.COMPILE_LATEX),
        ]
        if self.generate_cover_letter:
            self.steps.extend(
                [
                    StepProgress(step=TaskStep.EXTRACT_TEXT),
                    StepProgress(step=TaskStep.GENERATE_COVER_LETTER),
                    StepProgress(step=TaskStep.CREATE_COVER_PDF),
                ]
            )


class TaskCreate(BaseModel):
    job_description: str = ""
    generate_cover_letter: bool = True
    template_id: str = "classic"
    language: str = "en"
    provider: str | None = None


class TaskUpdate(BaseModel):
    job_description: str | None = None
    generate_cover_letter: bool | None = None
    template_id: str | None = None
    language: str | None = None
    provider: str | None = None


class TaskProgressUpdate(BaseModel):
    task_id: str
    step: TaskStep
    status: TaskStatus
    message: str = ""
    attempt: int = 0
